source/docs/advanced/graphql/slim-down-resolvers-with-loaders.md

```diff
@@ -0,0 +1,351 @@
+const counters = new Map();
+
+const currentValueOf = sku => {
+  return new Promise(resolve => {
+    setTimeout(() => {
+      const currentValue = counters.get(sku) || 0;
+      resolve(currentValue);
+    }, Math.random() * 3000);
+  });
+};
+
+const incrementValueOf = (sku, increment) => {
+  return currentValueOf(sku).then(currentValue =>
+    counters.set(sku, currentValue + increment)
+  );
+};
+
+export default {
+  loadBySku: sku => currentValueOf(sku),
+  incrementBySku: (sku, increment) => incrementValueOf(sku, increment)
+};
+```
+
+You can then update the resolvers to use this loader:
+
+```diff
+// my-module/server/modules/clicks-counters/resolvers.js
+-const counters = new Map();
++import CounterLoader from "./loader";
+
+-const currentValueOf = sku => {
+-  return new Promise(resolve => {
+-    setTimeout(() => {
+-      const currentValue = counters.get(sku) || 0;
+-      resolve(currentValue);
+-    }, Math.random() * 3000);
+-  });
+-};
+-
+module.exports = {
+  Product: {
+-    clicksCounter: ({ sku }) => currentValueOf(sku)
++    clicksCounter: ({ sku }) => CounterLoader.loadBySku(sku)
+  },
+
+  Mutation: {
+    incrementProductCounter(_, { sku, incrementValue = 1 }) {
+-      return currentValueOf(sku)
+-        .then(currentValue => counters.set(sku, currentValue + incrementValue))
++      return CounterLoader.incrementBySku(sku, incrementValue)
+        .then(() => ({
+          success: true
+        }));
+    }
+  }
+};
+```
+
+### Using GraphQL context for Dependency Injection
+
+In its current state, resolvers are tightly coupled to the current
+implementation. We recommend to leverage GraphQL’s context to inject a loader to
+resolvers.
+
+The GraphQL context is (according to
+[GraphQL.org](https://graphql.org/learn/execution/#root-fields-resolvers)):
+
+> A value which is provided to every resolver and holds important contextual
+> information like the currently logged in user, or access to a database.
+
+In Front-Commerce (and most GraphQL implementations), the `context` is passed as
+the 3rd argument of a resolver function.
+
+Let’s refactor resolvers to use a loader from the context instead of importing
+it:
+
+```diff
+// my-module/server/modules/clicks-counters/resolvers.js
+-import CounterLoader from "./loader";
+-
+module.exports = {
+  Product: {
+-    clicksCounter: ({ sku }) => CounterLoader.loadBySku(sku)
++    clicksCounter: ({ sku }, _, { loaders }) => {
++      return loaders.Counter.loadBySku(sku);
++    }
+  },
+
+  Mutation: {
+-    incrementProductCounter(_, { sku, incrementValue = 1 }) {
++    incrementProductCounter(_, { sku, incrementValue = 1 }, { loaders }) {
+-      return CounterLoader.incrementBySku(sku, incrementValue)
++      return loaders.Counter.incrementBySku(sku, incrementValue)
+        .then(() => ({
+          success: true
+        }));
+    }
+  }
+};
+```
+
+**You might be wondering why we used the `loaders` value from the context, and
```

Comment:
I'd put the contextEnhancer before this refactoring. Because this paragraph looks like a... doc smell :thinking: 

I mean that by this that it's not explicit why would people be wondering this. Even though it seems clear to us.

Reply:
I disagree on this one, even though I am not convinced either by the transition.

IMO it is really important to point out how the direct import in resolvers (that were just created above) creates coupling. Also I feel that developers will have a basic knowledge of DI and might see a point in this refactoring, while knowing that there is a « kind of DI container somewhere » doing some dark magic from an obscure YAML or XML configuration file.

I’ll keep it for now, but let’s change it if we see some incomprehension from readers…